### ISSUE TITLE
testing and chip targets?

### ISSUE THREAD

**Issue by kruus** [01org/mkl-dnn#136]
Any interest in supporting compilation flags that mirror a few `cpu_isa_t` types and limit jit support to the desired level?
- Example: -DTARGET_SSE42 would not support any `cpu_isa_t > `isa_avx2 or avx512*
  - so you could run `tests/` and `tests/gtests/` with max sse42 impls on your more powerful desktop.
- Example: -DTARGET_ANY would remove jit support
  - so you could run on an `isa_any` cpu (although slowly) (This is my interest)

Basically, supported memory formats/primitives/tests would be sorted and grouped by cpu class "everywhere", and I'd look into doing this without destroying readability too much.

**Comment by emfomenk** [comment]
At some point we will definitely need some sort of dispatch control in MKL-DNN.

But I am not sure that compile time limitation makes a lot of sense (maybe @rsdubtso has another opinion on that). Removing jit at all doesn't make a lot of sense for MLK-DNN (since optimizations are almost the only reason why MKL-DNN exists :)), except for debuging... maybe.

Summoning @rsdubtso...

**Comment by kruus** [comment]
OK, let's give it  a few days for others to weigh in.  I'm not particularly wedded to this, because it is extra work to do properly, and I'm happy enough just maintaining a fork and adding other primitive impls there.

#### run- or compile-time?
Run-time dispatch I used to think of mostly as some way to auto-tune between available impls, where choice is made about "fastest" (not what cpu impl runs on)
Run-time cpu limits might even be easier to support than compile time, now that I think of it.
- a primitive would simply have a method to return its "required cpu type".
- That's it. Test framework can iterate using this info as a filter.  Cool. Elegant.

#### An alternative view: jit is an *added bonus*.
After reviewing many platforms and libraries, we chose to adopting mkl-dnn **primarily for its API**.
- mkl-dnn is open-source
  - nvidia libs are not, and api did not really impress me (somewhat limited)
- It's from Intel, with a relation to Intel MKL APIs that are already used in many HPC products.
  - expect API to have good [and long-term] support, not just *the next craze*.
- the API focuses on performance-related layers in particular, so it is a **small** subset of any particular "framework".
  - By choosing to implement just the mkl-dnn API, we do not need to paint ourself in the corner of "we only Caffe/Tensorflow/Torch/Pytorch/Daal/mxnet/chainer/...&lt;insert your favorite&gt;...". Porting a low-level API is much easier than having a full ML platform running in some other OS or other CPU.
  - The layers chosen in mkl-dnn are the ones found to be particular performance sensitive.
- The layer APIs in mkl-dnn tend to support a *superset* of features found in any one platform.  mkl-dnn layers are *flexible* (OK for all nets).

Yes, I realize that jit support is a competitive necessity for Intel, but a good API and good open-source code are *very* strong features in their own right.
